test/malli/destructure_test.cljc

```diff
@@ -0,0 +1,316 @@
+(ns malli.destructure-test
+  (:require [clojure.test :refer [deftest testing is]]
+            [malli.destructure :as md]
+            [malli.core :as m]))
+
+(def expectations
+  [{:name "empty"
+    :bind '[]
+    :schema :cat}
+   {:name "1 arg"
+    :bind '[a]
+    :schema [:cat :any]}
+   {:name "2 args"
+    :bind '[a b]
+    :schema [:cat :any :any]}
+   {:name "2 + varargs"
+    :bind '[a b & cs]
+    :schema [:cat :any :any [:* :any]]}
+   {:name "sequence destructuring"
+    :bind '[a [b1 [b2] & bs :as bss] & [c1 c2 & cs :as css]]
+    :schema [:cat
+             :any
+             [:maybe
+              [:cat
+               [:? :any]
+               [:maybe
+                [:cat
+                 [:? :any]
+                 [:* :any]]]
+               [:* :any]]]
+             [:maybe
+              [:cat
+               [:? :any]
+               [:? :any]
+               [:* :any]]]]}
+   {:name "map destructuring"
+    :bind '[a {:keys [b]
+               :strs [c]
+               :syms [d]
+               :demo/syms [e]
+               :demo/keys [f]
+               g :demo/g
+               h 123
+               :or {b 0, d 0, f 0}
+               :as map}]
+    :schema [:cat
+             :any
+             [:altn
+              [:map [:map
+                     [:b {:optional true} :any]
+                     ["c" {:optional true} :any]
+                     ['d {:optional true} :any]
+                     ['demo/e {:optional true} :any]
+                     [:demo/f {:optional true}]
+                     [:demo/g {:optional true}]
+                     [123 {:optional true} :any]]]
```

Comment:
Hmm, okay. The key names are the keys from the map instead of the symbols they are being bound into. Makes sense now that I think about this, as the schema should describe the value.

Reply:
yes, the new `MapLike` schema groups the keys now by behaviour, so can't group by key anymore. 

Related, one can bind same symbol many times, making the binding... weird:

```clj
(let [{[a a] :a
       [b a] :b
       [c a] :c
       [d a] :d
       [e a] :d
       [f a] :d
       [g a] :d
       [h a] :d}
      {:a [1 2]
       :b [3 4]
       :c [5 6]
       :d [7 8]
       :e [9 10]
       :f [11 12]
       :g [13 14]
       :h [15 16]}] a)
; => ???
```

Reply:
If you meant to use the same key `:d` for all of the right hand side, then it's 8. If you meant to use each of the keys, then it's 16. They're assigned in the order written because destructuring happens at macro expansion time.

Reply:
oh, there was a bug in the code, too many `:d`s. Here's a example that highlights that this looks like undefined behavior (the map changes the impl, where we lose the key order:

```clj
(let [{[a a] :a
       [b a] :b
       [c a] :c
       [d a] :d
       [e a] :e
       [f a] :f
       [g a] :g
       [h a] :h
       [i a] :i}
      {:a [1 2]
       :b [3 4]
       :c [5 6]
       :d [7 8]
       :e [9 10]
       :f [11 12]
       :g [13 14]
       :h [15 16]
       :i [16 17]}] a)
; => 8


(let [{[a a] :a
       [b a] :b
       [c a] :c
       [d a] :d
       [e a] :e
       [f a] :f
       [g a] :g
       [h a] :h
       #_#_[i a] :i}
      {:a [1 2]
       :b [3 4]
       :c [5 6]
       :d [7 8]
       :e [9 10]
       :f [11 12]
       :g [13 14]
       :h [15 16]
       :i [16 17]}] a)
; => 16

(class '{[a a] :a
         [b a] :b
         [c a] :c
         [d a] :d
         [e a] :e
         [f a] :f
         [g a] :g
         [h a] :h
         #_#_[i a] :i})
; => clojure.lang.PersistentArrayMap

(class '{[a a] :a
         [b a] :b
         [c a] :c
         [d a] :d
         [e a] :e
         [f a] :f
         [g a] :g
         [h a] :h
         [i a] :i})
; => clojure.lang.PersistentHashMap
```